src/dune_pkg/resolve_opam_formula.ml

```diff
@@ -79,6 +79,19 @@ end
 type unsatisfied_formula =
   [ `Formula_could_not_be_satisfied of Unsatisfied_formula_hint.t list ]
 
+let dedup_package_names package_names =
+  let ret_reversed, _ =
+    List.fold_left
+      package_names
+      ~init:([], Package_name.Set.empty)
+      ~f:(fun (ret_reversed, seen) x ->
+        if Package_name.Set.mem seen x
+        then ret_reversed, seen
+        else x :: ret_reversed, Package_name.Set.add seen x)
+  in
+  List.rev ret_reversed
+;;
```

Comment:
It could potentially make sense to have a

```
val uniq : equal:('a -> 'a -> bool) -> 'a list -> 'a t
```

style function in `Stdune.List`, but this is fine too.

Reply:
Ah, I just noticed that `List.sort_uniq` already exists. It will obviously reshuffle the order, but might be ok for the usecase.

Reply:
Note that we aren't allowed to re-order the dependencies as specified by the user. The order has a semantic meaning.

Reply:
In the case of `libraries` yes, but do OPAM dependencies have a semantic ordering?

Reply:
I think we should preserve the order of dependencies in lockfiles if for no other reason than to simplify the mental model for the correspondence between lockfiles and the opam files they are generated from. If some fields are copied but reordered then it's more work to understand how the opam file maps to the lockfile.